iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/read/filescan/impl/DiskChunkHandleImpl.java

```diff
@@ -0,0 +1,118 @@
+
+import org.apache.tsfile.file.header.ChunkHeader;
+import org.apache.tsfile.file.header.PageHeader;
+import org.apache.tsfile.file.metadata.statistics.Statistics;
+import org.apache.tsfile.read.TsFileSequenceReader;
+import org.apache.tsfile.read.common.Chunk;
+import org.apache.tsfile.read.reader.chunk.ChunkReader;
+import org.apache.tsfile.utils.ReadWriteForEncodingUtils;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.nio.ByteBuffer;
+
+/** It will receive a list of offset and execute sequential scan of TsFile for chunkData. */
+public class DiskChunkHandleImpl implements IChunkHandle {
+
+  protected ChunkHeader currentChunkHeader;
+  protected PageHeader currentPageHeader;
+  protected ByteBuffer currentChunkDataBuffer;
+  protected TsFileSequenceReader reader;
+  protected long offset;
+
+  // Page will reuse chunkStatistics if there is only one page in chunk
+  protected final Statistics<? extends Serializable> chunkStatistic;
+
+  public DiskChunkHandleImpl(
+      TsFileSequenceReader reader,
+      long offset,
+      Statistics<? extends Serializable> chunkStatistics) {
+    this.chunkStatistic = chunkStatistics;
+    this.reader = reader;
+    this.offset = offset;
+  }
+
+  protected void init() throws IOException {
+    if (currentChunkDataBuffer != null) {
+      return;
+    }
+    Chunk chunk = reader.readMemChunk(offset);
+    this.currentChunkDataBuffer = chunk.getData();
+    this.currentChunkHeader = chunk.getHeader();
+  }
+
+  // Check if there is more pages to be scanned in Chunk.
+  // If so, deserialize the page header
+  @Override
+  public boolean hasNextPage() throws IOException {
+    // read chunk from disk if needed
+    init();
+    if (!currentChunkDataBuffer.hasRemaining()) {
+      return false;
+    }
+    // If there is only one page, page statistics is not stored in the chunk header, which is the
+    // same as chunkStatistics
+    if ((byte) (this.currentChunkHeader.getChunkType() & 63) == 5) {
+      currentPageHeader =
+          PageHeader.deserializeFrom(this.currentChunkDataBuffer, this.chunkStatistic);
+    } else {
+      currentPageHeader =
+          PageHeader.deserializeFrom(
+              this.currentChunkDataBuffer, this.currentChunkHeader.getDataType());
+    }
+    return true;
+  }
+
+  @Override
+  public void skipCurrentPage() {
+    currentChunkDataBuffer.position(
+        currentChunkDataBuffer.position() + currentPageHeader.getCompressedSize());
+  }
+
+  @Override
+  public long[] getPageStatisticsTime() {
+    return new long[] {currentPageHeader.getStartTime(), currentPageHeader.getEndTime()};
+  }
+
+  @Override
+  public long[] getDataTime() throws IOException {
+    ByteBuffer currentPageDataBuffer =
+        ChunkReader.deserializePageData(
+            currentPageHeader, this.currentChunkDataBuffer, this.currentChunkHeader);
+    int timeBufferLength = ReadWriteForEncodingUtils.readUnsignedVarInt(currentPageDataBuffer);
```

Comment:
timeBufferLength doesn't equal to `line count * 8` because `timeBufferLength` is an encoded bin stream.

Reply:
you can get the time length from page header.